### PR TITLE
 [pytorch][build][sagemaker] PyTorch 1.12.1 SageMaker release

### PR DESCRIPTION
**Pull request by ShiboXing**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Sagemaker images release for pytorch 1.12.1
### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [x] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.12/py3/cu116/Dockerfile.gpu [hunk 1]
```diff
@@ -4,11 +4,11 @@ ARG PYTHON_SHORT_VERSION=3.8
 ARG MAMBA_VERSION=4.12.0-2
```
Comment (ydaiming) [inline pytorch/training/docker/1.12/py3/cu116/Dockerfile.gpu:4]: Why do you change this cu116 dockerfile? It's only cu113 for SM.

File: pytorch/inference/docker/1.12/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -3,10 +3,10 @@ ARG PYTHON_VERSION=3.8.13
 ARG MAMBA_VERSION=4.12.0-2
 
 # PyTorch Binaries
-ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_e3/20220714-051536/64a46e47aea7de3d99f75c5b4feba6d66c80af8b/torch-1.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cpu/torchvision-0.13.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cpu/torchaudio-0.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_e3/20220825-170010/c0e08ff275dba462c0807c8ccaef2584bb7003a4//codebuild/output/src087/src/s3/01/torch-1.12.1%2Bcpu-cp38-cp38-linux_x86_64.whl
```
Comment (ydaiming) [inline pytorch/inference/docker/1.12/py3/Dockerfile.cpu:6]: Could you manually update the path to remove the unnecessary `/codebuild/output/src087/src/s3/01/` portion?

File: pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 3]
```diff
@@ -2,10 +2,10 @@ ARG PYTHON=python3
 ARG PYTHON_VERSION=3.8.13
 
 # PyTorch Binaries
-ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_e3/20220714-012926/64a46e47aea7de3d99f75c5b4feba6d66c80af8b/torch-1.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cu113/torchvision-0.13.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cu113/torchaudio-0.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_e3/20220825-173005/c0e08ff275dba462c0807c8ccaef2584bb7003a4//codebuild/output/src022/src/s3/01/torch-1.12.1%2Bcu113-cp38-cp38-linux_x86_64.whl
```
Comment (ydaiming) [inline pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu:5]: Same as above

File: pytorch/training/docker/1.12/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -3,11 +3,11 @@ ARG PYTHON_VERSION=3.8.13
 ARG MAMBA_VERSION=4.12.0-2
 
 # PyTorch Binaries
-ARG PT_EC2_TRAINING_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_e3/20220714-051536/64a46e47aea7de3d99f75c5b4feba6d66c80af8b/torch-1.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_SM_TRAINING_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_sm/20220713-224008/9f169f548f4b1eca6631c07d48bbe73a59e300ca/torch-1.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cpu/torchvision-0.13.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cpu/torchaudio-0.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_EC2_TRAINING_URL=https://download.pytorch.org/whl/cpu/torch-1.12.1%2Bcpu-cp38-cp38-linux_x86_64.whl
+ARG PT_SM_TRAINING_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_sm/20220825-171049/4c2f0ae5d3095b8d484fc181be42af2705173d59//codebuild/output/src254/src/s3/01/torch-1.12.1%2Bcpu-cp38-cp38-linux_x86_64.whl
```
Comment (ydaiming) [inline pytorch/training/docker/1.12/py3/Dockerfile.cpu:7]: Same as above

File: pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 5]
```diff
@@ -4,11 +4,11 @@ ARG PYTHON_SHORT_VERSION=3.8
 ARG MAMBA_VERSION=4.12.0-2
 
 # PyTorch Binaries
-ARG PT_EC2_TRAINING_URL=https://download.pytorch.org/whl/test/cu113/torch-1.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_SM_TRAINING_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_sm/20220714-012958/9f169f548f4b1eca6631c07d48bbe73a59e300ca/torch-1.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cu113/torchvision-0.13.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cu113/torchaudio-0.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_EC2_TRAINING_URL=https://download.pytorch.org/whl/cu113/torch-1.12.1%2Bcu113-cp38-cp38-linux_x86_64.whl
+ARG PT_SM_TRAINING_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_sm/20220825-173356/4c2f0ae5d3095b8d484fc181be42af2705173d59//codebuild/output/src121/src/s3/01/torch-1.12.1%2Bcu113-cp38-cp38-linux_x86_64.whl
```
Comment (ydaiming) [inline pytorch/training/docker/1.12/py3/cu113/Dockerfile.gpu:8]: Same as above

**Review by ydaiming — CHANGES_REQUESTED** [review]

**Review by ydaiming — APPROVED** [review]
Looks good if build and tests pass

File: pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 6]
```diff
@@ -2,10 +2,10 @@ ARG PYTHON=python3
 ARG PYTHON_VERSION=3.8.13
 
 # PyTorch Binaries
-ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_e3/20220714-012926/64a46e47aea7de3d99f75c5b4feba6d66c80af8b/torch-1.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cu113/torchvision-0.13.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cu113/torchaudio-0.12.0%2Bcu113-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_e3/20220826-043953/579ffd0893a9d3e60773918bef0b000dfe60a7ce/torch-1.12.1%2Bcu113-cp38-cp38-linux_x86_64.whl
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu:5]: please remove the "e3" and change with "ec2"

File: pytorch/inference/docker/1.12/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -3,10 +3,10 @@ ARG PYTHON_VERSION=3.8.13
 ARG MAMBA_VERSION=4.12.0-2
 
 # PyTorch Binaries
-ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_e3/20220714-051536/64a46e47aea7de3d99f75c5b4feba6d66c80af8b/torch-1.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHVISION_URL=https://download.pytorch.org/whl/cpu/torchvision-0.13.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHAUDIO_URL=https://download.pytorch.org/whl/cpu/torchaudio-0.12.0%2Bcpu-cp38-cp38-linux_x86_64.whl
-ARG PT_TORCHDATA_URL=https://download.pytorch.org/whl/test/torchdata-0.4.0-cp38-cp38-linux_x86_64.whl
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.1_e3/20220826-041141/579ffd0893a9d3e60773918bef0b000dfe60a7ce/torch-1.12.1%2Bcpu-cp38-cp38-linux_x86_64.whl
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/Dockerfile.cpu:6]: please remove the "e3" and change with "ec2"

**Review by arjkesh — COMMENTED** [review]
Were any sagemaker tests run during the development of this PR? If so, can you point to the commit hash so I can verify?

Please fill out the necessary information in the PR description body

**Review by roywei — CHANGES_REQUESTED** [review]
SMDDP needs to be re-compiled with this new wheel. The smddp test should be failing for this PR. Could you double check?
I'm running the image locally on 2 p4d and failed when training mnist.
```
12: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
13: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
14: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
15: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
 8: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
 9: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
10: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
11: python: /smddp/thirdparty/pybind11/include/pybind11/pybind11.h:1353: void pybind11::detail::generic_type::initialize(const pybind11::detail::type_record&):
 Assertion `parent_tinfo != nullptr' failed.
```


Please use this new wheel and verify tests are passing, thanks!

https://smdataparallel.s3.amazonaws.com/binary/pytorch/1.12.1/cu113/2022-08-31/smdistributed_dataparallel-1.5.1-cp38-cp38-linux_x86_64.whl

**Review by jeet4320 — APPROVED** [review]
need security tests to be run. Other tests are passing

**Comment by jeet4320** [comment]
can you enable below tests and run?

```
  safety_check_test = false
  ecr_scan_allowlist_feature = false
```

**Review by jeet4320 — APPROVED** [review]

**Review by tejaschumbalkar — DISMISSED** [review]

File: pytorch/buildspec.yml [hunk 8]
```diff
@@ -46,139 +46,139 @@ context:
       target: deep_learning_container.py
 
 images:
```
Comment (tejaschumbalkar) [inline pytorch/buildspec.yml:48]: revert changes to this file

File: dlc_developer_config.toml [hunk 9]
```diff
@@ -33,22 +33,22 @@ do_build = true
 [test]
 ### On by default
 sanity_tests = true
-  safety_check_test = false
-  ecr_scan_allowlist_feature = false
+  safety_check_test = true
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:36]: revert changes to this file

File: dlc_developer_config.toml [hunk 10]
```diff
@@ -99,4 +99,4 @@ dlc-pr-tensorflow-2-graviton = ""
 dlc-pr-mxnet-eia = ""
 dlc-pr-pytorch-eia = ""
 dlc-pr-tensorflow-1-eia = ""
-dlc-pr-tensorflow-2-eia = ""
+dlc-pr-tensorflow-2-eia = ""
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:102]: add a next line

File: pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu [hunk 11]
```diff
@@ -2,10 +2,10 @@ ARG PYTHON=python3
 ARG PYTHON_VERSION=3.8.13
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu:2]: I dont see binary changes for inference cpu dockerfiles

**Review by tejaschumbalkar — COMMENTED** [review]
Please uncomment SM build configuration in `buildpsec.yml` file

File: pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu (hunk 11 above)
Comment (ShiboXing) [reply pytorch/inference/docker/1.12/py3/cu113/Dockerfile.gpu:2]: CPU inference binary change has been merged into master by the 1.12.1_ec2 PR

File: pytorch/buildspec.yml [hunk 12]
```diff
@@ -76,52 +76,52 @@ images:
...
+                         *DEVICE_TYPE ]
+    target: sagemaker
+    context:
+      <<: *TRAINING_CONTEXT
+  BuildPyTorchExampleGPUTrainPy3DockerImage:
+    <<: *TRAINING_REPOSITORY
+    build: &PYTORCH_GPU_TRAINING_PY3 false
+    # enable_test_promotion: false
+    image_size_baseline: 14000
+    #base_image_name: BuildEC2GPUPTTrainPy3DockerImage
```
Comment (tejaschumbalkar) [inline pytorch/buildspec.yml:114]: uncomment this line

**Review by tejaschumbalkar — APPROVED** [review]

**Pull request merged**